Compiling issue with GAPI

Hi,

I try to compile the last version of the code under osx.

Here the cmake command:

`cmake -DWITH_QT=ON -DWITH_OPENGL=ON -DWITH_TBB=ON -DBUILD_EXAMPLES=OFF -DBUILD_DOCS=OFF -DBUILD_TESTS=OFF -OPENCV_CONTRIB=OFF -DENABLE_PRECOMPILED_HEADERS=OFF -DOPENCV_GENERATE_PKGCONFIG=ON -DCMAKE_PREFIX_PATH=/Applications/Qt/5.12.3/clang_64/ ..`

here the error that I get:

```
In file included from /Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/compiler/gcompiler.cpp:31:
In file included from /Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/compiler/gstreaming_priv.hpp:12:
/Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/executor/gstreamingexecutor.hpp:19:12: fatal error: 'tbb/concurrent_queue.h' file not found
#  include <tbb/concurrent_queue.h> // FIXME: drop it from here!
```

Looks like that the path for TBB is missing.

I have the TBB installed here: /usr/local/include/tbb

```
[ 73%] Building CXX object modules/gapi/CMakeFiles/opencv_gapi.dir/src/compiler/gcompiler.cpp.o
cd /Users/thewoz/Documents/GitHub/opencv/build/modules/gapi && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/thewoz/Documents/GitHub/opencv/build -I/Users/thewoz/Documents/GitHub/opencv/modules/gapi/include -I/Users/thewoz/Documents/GitHub/opencv/build/modules/gapi -I/Users/thewoz/Documents/GitHub/opencv/modules/core/include -I/Users/thewoz/Documents/GitHub/opencv/modules/imgproc/include -I/Users/thewoz/Documents/GitHub/opencv/modules/gapi/src -I/Users/thewoz/Documents/GitHub/opencv/build/3rdparty/ade/ade-0.1.1f/sources/ade/include -isystem /usr/local/include/eigen3  -fsigned-char -ffast-math -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -fdiagnostics-show-option -Wno-long-long -Qunused-arguments -Wno-semicolon-before-method-body -ffunction-sections -fdata-sections  -fvisibility=hidden -fvisibility-inlines-hidden -g  -O0 -DDEBUG -D_DEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC   -std=c++11 -o CMakeFiles/opencv_gapi.dir/src/compiler/gcompiler.cpp.o -c /Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/compiler/gcompiler.cpp
In file included from /Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/compiler/gcompiler.cpp:31:
In file included from /Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/compiler/gstreaming_priv.hpp:12:
/Users/thewoz/Documents/GitHub/opencv/modules/gapi/src/executor/gstreamingexecutor.hpp:19:12: fatal error: 
      'tbb/concurrent_queue.h' file not found
#  include <tbb/concurrent_queue.h> // FIXME: drop it from here!

```

--- a/cmake/OpenCVUtils.cmake
+++ b/cmake/OpenCVUtils.cmake
@@ -1412,7 +1412,9 @@ function(ocv_target_link_libraries target)
   foreach(dep ${LINK_DEPS})
     if(" ${dep}" STREQUAL " ${target}")
       # prevent "link to itself" warning (world problem)
-    elseif(" ${dep}" STREQUAL " LINK_PRIVATE" OR " ${dep}" STREQUAL "LINK_PUBLIC")
+    elseif(" ${dep}" STREQUAL " LINK_PRIVATE" OR " ${dep}" STREQUAL "LINK_PUBLIC"
+        OR " ${dep}" STREQUAL " PRIVATE" OR " ${dep}" STREQUAL "PUBLIC"
+    )
       if(NOT LINK_PENDING STREQUAL "")
         __ocv_push_target_link_libraries(${LINK_MODE} ${LINK_PENDING})
         set(LINK_PENDING "")

--- a/modules/gapi/CMakeLists.txt
+++ b/modules/gapi/CMakeLists.txt
@@ -1,15 +1,13 @@
 # FIXME: Rework standalone build in more generic maner
 # (Restructure directories, add common pass, etc)
-if (NOT DEFINED OPENCV_INITIAL_PASS)
+if(NOT DEFINED OPENCV_INITIAL_PASS)
     cmake_minimum_required(VERSION 3.3)
     project(gapi_standalone)
     include("cmake/standalone.cmake")
     return()
 endif()
 
-# FIXME: Remove CXX11 check after complete switch to OpenCV 4 branch
-# (CI, bundle, workloads, etc)
-if (NOT HAVE_CXX11 OR NOT TARGET ade)
+if(NOT TARGET ade)
   # can't build G-API because of the above reasons
   ocv_module_disable(gapi)
   return()
@@ -131,9 +129,13 @@ ocv_source_group("Include" FILES ${gapi_ext_hdrs})
 ocv_set_module_sources(HEADERS ${gapi_ext_hdrs} SOURCES ${gapi_srcs})
 ocv_module_include_directories("${CMAKE_CURRENT_LIST_DIR}/src")
 
-# Note `ade` is not a module name but link dependency for ${the_module}
-# (which is opencv_gapi)
-ocv_create_module(ade ${INF_ENGINE_TARGET})
+ocv_create_module()
+
+ocv_target_link_libraries(${the_module} PRIVATE ade ${INF_ENGINE_TARGET})
+if(HAVE_TBB)
+  ocv_target_link_libraries(${the_module} PRIVATE tbb)
+endif()
+
 ocv_add_accuracy_tests(${INF_ENGINE_TARGET})
 # FIXME: test binary is linked with ADE directly since ADE symbols
 # are not exported from libopencv_gapi.so in any form - thus
@@ -144,17 +146,17 @@ if(TARGET opencv_test_gapi)
   target_link_libraries(opencv_test_gapi PRIVATE ade)
 endif()
 
-if (HAVE_FREETYPE)
+if(HAVE_FREETYPE)
   ocv_target_compile_definitions(opencv_gapi      PRIVATE -DHAVE_FREETYPE)
   ocv_target_compile_definitions(opencv_test_gapi PRIVATE -DHAVE_FREETYPE)
-  ocv_target_link_libraries(opencv_gapi LINK_PRIVATE ${FREETYPE_LIBRARIES})
+  ocv_target_link_libraries(opencv_gapi PRIVATE ${FREETYPE_LIBRARIES})
   ocv_target_include_directories(opencv_gapi PRIVATE ${FREETYPE_INCLUDE_DIRS})
 endif()
 
 if(HAVE_PLAIDML)
   ocv_target_compile_definitions(opencv_gapi      PRIVATE -DHAVE_PLAIDML)
   ocv_target_compile_definitions(opencv_test_gapi PRIVATE -DHAVE_PLAIDML)
-  ocv_target_link_libraries(opencv_gapi LINK_PRIVATE ${PLAIDML_LIBRARIES})
+  ocv_target_link_libraries(opencv_gapi PRIVATE ${PLAIDML_LIBRARIES})
   ocv_target_include_directories(opencv_gapi SYSTEM PRIVATE ${PLAIDML_INCLUDE_DIRS})
 endif()
 

